packages/codeeditor/src/editor.ts

```diff
@@ -227,123 +234,81 @@ export namespace CodeEditor {
     /**
      * Construct a new Model.
      */
-    constructor(options?: Model.IOptions) {
-      options = options || {};
-      if (options.modelDB) {
-        this.modelDB = options.modelDB;
+    constructor(options: Model.IOptions) {
+      this._isDocument = options.isDocument;
+
+      if (this._isDocument) {
+        // The IModel is a document, so we create the
+        // `source` from the root of the ISharedDoc
+        this._sharedDoc = options.sharedDoc || new SharedDoc();
+        this._value = this._sharedDoc.createString('source') as SharedString;
+        this._value.undoManager = SharedDoc.createUndoManager(
+          (this._value as SharedString).underlyingModel,
+          []
+        );
+        this._value.changed.connect(this._onValueChanged, this);
+        this._triggerModelReady();
       } else {
-        this.modelDB = new ModelDB();
+        this._sharedDoc = options.sharedDoc!;
       }
-      this.sharedModel = models.createStandaloneCell(
-        this.type,
-        options.id
-      ) as models.ISharedText;
-      this.sharedModel.changed.connect(this._onSharedModelChanged, this);
-
-      const value = this.modelDB.createString('value');
-      value.changed.connect(this._onModelDBValueChanged, this);
-      value.text = value.text || options.value || '';
 
-      const mimeType = this.modelDB.createValue('mimeType');
-      mimeType.changed.connect(this._onModelDBMimeTypeChanged, this);
-      mimeType.set(options.mimeType || 'text/plain');
-
-      this.modelDB.createMap('selections');
+      this._mimeType = options.mimeType || 'text/plain';
+      this._selections = new ObservableMap<ITextSelection[]>();
     }
 
     /**
-     * When we initialize a cell model, we create a standalone model that cannot be shared in a YNotebook.
-     * Call this function to re-initialize the local representation based on a fresh shared model (e.g. models.YFile or models.YCodeCell).
-     *
-     * @param sharedModel
-     * @param reinitialize Whether to reinitialize the shared model.
+     * The type of the IModel
      */
-    public switchSharedModel(
-      sharedModel: models.ISharedText,
-      reinitialize?: boolean
-    ): void {
-      if (reinitialize) {
-        // update local modeldb
-        // @todo also change metadata
-        this.value.text = sharedModel.getSource();
-      }
-      this.sharedModel.changed.disconnect(this._onSharedModelChanged, this);
-      // clone model retrieve a shared (not standalone) model
-      this.sharedModel = sharedModel;
-      this.sharedModel.changed.connect(this._onSharedModelChanged, this);
-      this._sharedModelSwitched.emit(true);
+    get type(): nbformat.CellType {
```

Comment:
From @afshin: We probably should not depend on `nbformat` in this place. Do we need the `type` method here.